### PR TITLE
Feature/mint and call

### PR DESCRIPTION
**Pull request by malteish**
Testing first specification of `mintAndCall` for usefulness.

Run the first test like this:
```bash
forge test --match-path test/BuyWithMintAndCall.t.sol 
```

### CODE REVIEW

File: contracts/Wallet.sol [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+
+    constructor(ContinuousFundraising _fundraising) Ownable2Step() {
+        fundraising = _fundraising;
+    }
+
+    /**
+    @notice sets (or updates) the receiving address for an IBAN hash
+    */
+    function set(bytes32 _ibanHash, address _tokenReceiver) external onlyOwner {
+        receiverAddress[_ibanHash] = _tokenReceiver;
```
Comment (gislik) [inline contracts/Wallet.sol:33]: That's exactly the way we envisioned you using the hashed IBAN 🙌

File: contracts/Wallet.sol [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+            return bytes4(0xDEADD00D); // ERC1363ReceiverNotRegistered
+        }
+        address tokenReceiver = receiverAddress[ibanHash];
+        // todo: calculate amount
+        uint256 amount = 100;
+        // grant allowance to fundraising
+        IERC20(fundraising.currency()).approve(address(fundraising), amount);
+        // todo: add try catch https://solidity-by-example.org/try-catch/
+        fundraising.buy(amount, tokenReceiver);
+        return 0x600D600D; // ERC1363ReceiverSuccess
```
Comment (gislik) [inline contracts/Wallet.sol:58]: Related to @CJentzsch comment above ERC677 expects a bool to be returned.

File: contracts/Wallet.sol [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+
+    /*
+    @notice withdraws tokens to a given address
+    */
+    function withdraw(
+        address token,
+        address to,
+        uint256 amount
+    ) external onlyOwner {
+        IERC20(token).safeTransferFrom(address(this), to, amount);
```
Comment (gislik) [inline contracts/Wallet.sol:69]: Why not use `IERC20(token).safeTransfer(to, amount)`. The wallet is the owner of the tokens at this stage, right?

**Review by gislik — COMMENTED** [review]
lgtm 🚀

The only catch is that we rely on ERC677 and not ERC1363.

I know this is a POC, so e.g. the `withdraw` method is very simple. I have some ideas on how you can allow each investor to claim their "portion" of the funds in the wallet. To be discussed during our next call 😀

**Comment by malteish** [comment]
> The only catch is that we rely on ERC677 and not ERC1363.

[ERC677](https://github.com/ethereum/EIPs/issues/677) seems abandoned to me. The last comments on the (closed) issue indicate that is will not be continued, partially because [EIP1363](https://eips.ethereum.org/EIPS/eip-1363) solves all problems ERC677 addressed. EIP1363 is final. Sadly, it is the more complex solution, but it is a good solution non the less. I don't like to keep poking the dead ERC677 horse with a stick to make it move. 

I missed parts of the thought process that lead to the decision for ERC677 though. @CJentzsch, @gislik are there good reasons for ERC677 and against EIP1363 that I need to know?

File: contracts/Wallet.sol (hunk 2 above)
Comment (malteish) [reply contracts/Wallet.sol:58]: https://github.com/corpus-io/tokenize.it-smart-contracts/pull/138#issuecomment-1531036641

File: contracts/Wallet.sol (hunk 3 above)
Comment (malteish) [reply contracts/Wallet.sol]: true. [f3ef099](https://github.com/corpus-io/tokenize.it-smart-contracts/pull/138/commits/f3ef09939e4fba014b359602dd235762bf9651fd)

File: contracts/Wallet.sol [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+    function set(bytes32 _ibanHash, address _tokenReceiver) external onlyOwner {
+        receiverAddress[_ibanHash] = _tokenReceiver;
+        emit Set(_ibanHash, _tokenReceiver);
+    }
+
+    /**
+     * @notice ERC1363 callback
+     * @dev to support the mintAndCall standard, this function MUST NOT revert!
+     */
+    function onTransferReceived(
```
[the remark this answers is not in the document]
Comment (malteish) [reply contracts/Wallet.sol:41]: https://github.com/corpus-io/tokenize.it-smart-contracts/pull/138#issuecomment-1531036641

**Comment by malteish** [comment]
Currently, we are forced to trust the payment token contract (EURe). It is possible to spend funds in `Wallet` by calling `onTokenReceived()`. This has been mitigated by allowing only the payment token contract to call `onTokenReceived()`, which is not perfect.
See [this test](https://github.com/corpus-io/tokenize.it-smart-contracts/pull/138/files#diff-0376d8fdd666d7bc6af5119d5cf9f59d4306e59e413262206e053cbd050b23a2R127) for an example

**Comment by malteish** [comment]
we will not implement this now

**Pull request closed without merging**